### PR TITLE
Add FreestyleDynamic component to interact with component properties

### PR DESCRIPTION
**Pull request by danwenzel**
### Problem

For complex components with many inter-related properties, it becomes more and more cumbersome to have to write a variant for each possible state the component could be in.  If you want to cover all combinations of properties, you'd end up with 10s or even 100s of variants.

Also, designers and engineers are not able to edit these properties from within the component guide.  

### Solution

The solution is `freestyle-dynamic`.  First, build a `dynamicProperties` object in your javascript:

```js
dynamicProperties: {
  title: {
    value: 'Dynamic Title'
  },
  isTasteful: {
    value: false,
    inputType: 'checkbox',
    description: 'Changes to a tasteful color scheme'
  }
  ...
}
```

Then, call the `freestyle-dynamic` component, passing in your dynamic properties:
```hbs
{{freestyle-dynamic 'dynamic-properties'
  dynamicProperties=dynamicProperties
  title='Dynamic Properties'
  componentName='x-bar'
}}
```

The dynamic properties are passed to the component, and the values are bound to input fields that show up above the Freestyle usage area and are editable.  The code snippet is also dynamic, changing when you change the inputs.

### Demo

ember-freestyle-dynamic

### Notes

This is something we've been using in our organization for a while now.  It has become an essential tool for developers and designers to come up with custom designs for specs, and to test component changes against an infinite variety of property values.

### TODO
- Clean up styles
- Write tests
- Add documentation to README

### Nice-to-haves (future vision)
- More input types (currently only text, textarea, and checkbox) _UPDATE: number and select added!_
- JSON editor for properties that are objects
- ~~Have the code snippet dynamically change when the properties change~~ _DONE!_

### CODE REVIEW

**Comment by chrislopresto** [comment]
Thanks for this.. the approach looks really neat. I have wanted to figure out a reusable technique to wire up interactive, backing data for components in a freestyle guide for a while now. One way or another, I'd love to iterate on this idea and get something merged in.

How complex are the freestyle guides where you have used this technique?

**Comment by danwenzel** [comment]
Hey Chris, glad you're interested.  We have been using this for our style guide here at PagerDuty (we call it a "Component Guide").  Most of our front-end is Ember, and we currently have about 40 entries in the guide.  Of those, about 10 are using this dynamic properties technique.  Each entry itself varies in complexity, ranging from around 3 to 10 dynamic properties.

The guide continues to grow rapidly as our front-end evolves and our engineers and designers get more and more accustomed to using this tool.

I'm happy to answer any other questions you have and work with you to get this shipped.  Thanks!

**Comment by simonihmig** [comment]
For inspiration: recently I was attending a conference talk on this thing: http://react-styleguidist.js.org/

When you click on 'Show code' it allows you to play with the actual "template" code. Probably not so easy to implement with Ember, as you would have to ship the template compiler to the browser!?

**Comment by chrislopresto** [comment]
@danwenzel Nice! I want to shift over to "component guide" or "component explorer" terminology everywhere in Freestyle. 

Regarding the dynamic properties, I want to work towards an approach that somehow ties to the component props themselves. @lukemelia linked me to http://ciena-blueplanet.github.io/ember-prop-types/ the other day.

As you allude to @simonihmig, there's a lot more happening in this area in the React community. I'd love to take as much inspiration from projects like react-styleguidist as possible.

As far as the code preview, I do love how the freestyle-usage component hoists actual application code straight on into the freestyle component guide. But once we start stubbing dynamic data sources (we've done a lot of that as well, @danwenzel), the actual code preview starts to look a lot less like actual usage. We wouldn't have something called `dynamicProperties` in our actual app code.

So I wonder if we should take another cue from react-styleguidist and show the actual POJO values in the code preview. We could build a different component so that folks could still use freestyle-usage (names can all be changed) to show actual code.

Regardless of eventual implementation details, I'm looking forward to moving the project in this direction!

**Comment by danwenzel** [comment]
@chrislopresto - Sounds good!  I'll have some time this Friday to dive more into this.  Thanks for those references - that should definitely help.

**Comment by danwenzel** [comment]
@chrislopresto - I was able build a way to make the snippet dynamic.  It involves manually building the snippet from the dynamic properties.  The result:

image

The only drawback is that you still have to call the component itself along with passing extra arguments into `freestyle-dynamic`:

```hbs
{{#freestyle-section name='Dynamic Properties' as |section|}}
    {{#freestyle-dynamic 'dynamic-properties'
      dynamicProperties=dynamicProperties
      title='Dynamic Properties'
      componentName='x-bar'
    }}
      {{x-bar
        title=dynamicProperties.title.value
        description=dynamicProperties.description.value
        showBorder=dynamicProperties.showBorder.value
        isVisible=dynamicProperties.isVisible.value
        isTasteful=dynamicProperties.isTasteful.value
       }}
    {{/freestyle-dynamic}}
  {{/freestyle-section}}
```

It would be really great if you could dynamically build the component, but I haven't found a good way to do that, as you can't call a component with dynamic property names.

Let me know if you like the track I'm on, and I'll continue to clean up, add tests, etc.

**Comment by danwenzel** [comment]
As for using the component's actual properties, I do like the looks of `ember-prop-types`, and it would be cool to be able to hook into that.  One thing, though, is that I don't think we should _require_ the user to have to use this add-on.  Maybe we could give them the option of either manually passing `dynamicProperties` as I have currently, or using `ember-prop-types`?

**Comment by danwenzel** [comment]
Alright... I found a way to have `freestyle-dynamic` automatically render the component, setting the dynamic properties (see last 2 commits).  It's a bit hacky, as I had to reopen `Ember.Component` and add a new hook into `didReceiveAttrs`. But this should also make it easy to pull in `ember-prop-types` properties.

Again, let me know if I should continue down this track, or if you have other ideas.

**Comment by danwenzel** [comment]
@chrislopresto - Have you had a chance to take a look at this?

**Comment by simonihmig** [comment]
Would love to see this getting forward! @chrislopresto what's your thoughts on this?

Also agree that we should not require to use ember-prop-types here, but an optional integration would be nice though...

@danwenzel would like to play with your branch, even when not merged yet. Any chance you could rebase this, as there are some merge conflicts?

**Comment by chrislopresto** [comment]
@simonihmig @danwenzel Sorry for the delay. This is some really nifty work. I'm really excited to push forward in this direction and would like to home in on something mergeable that we can begin to iterate on.

I'd be happy to pair on a rebase.

**Comment by danwenzel** [comment]
Hey @chrislopresto and @simonihmig.  Just re-merged master, so this should be current now

**Comment by simonihmig** [comment]
@danwenzel thanks a lot! Currently on vacation, but will try this out once I am back at work...

**Comment by danwenzel** [comment]
Awesome!  And feel free, Chris and Simon, to review/leave comments on what I've got so far.  I should also have time to work more on this within a couple weeks.

**Comment by simonihmig** [comment]
@danwenzel I finally found the time now to try this out a bit. My first thoughts:

* the main input type we would need for most of our components would probably be a select box! Without that we would not be able to really make use of this
* it seems you cannot currently use a component in block form, right? This would be very limiting as well...
* there may be cases where you might not want to have all component inputs as dynamic form fields, because that would clutter the UI too much. So using a mix of static and dynamic props would be nice.

Don't get me wrong in light of this "criticism", it is absolutely awesome where this is heading to, and I am very thankful for your efforts here! 👍  But maybe we can still improve this a bit...

I have not really thought about this implementation wise, but from the perspective of a user, I would like to think a bit about this alternative API you posted [here](https://github.com/chrislopresto/ember-freestyle/pull/100#issuecomment-301260104). What if we could do something like...

```hbs
{{#freestyle-dynamic 'dynamic-properties'
  dynamicProperties=dynamicProperties
  title='Dynamic Properties'
  as |dynamic|
}}
  {{#x-bar
    title='Just a static title'
    showBorder=dynamic.showBorder
    isVisible=dynamic.isVisible
    isTasteful=dynamic.isTasteful
    options=model
  }}
    <p>{{dynamic.description}}</p>
  {{/x-bar}}
{{/freestyle-dynamic}}
```

This shows a mix of static and dynamic properties, as well as a dynamic one which however does not come from the dynamic properties form (`options=model`). An example for the latter could be something like a smart component, that expects an ember-data `RecordArray` of some type. That input cannot be static nor come from the form's `dynamicProperties`. 

Also the example above would allow for a block template, with a mix of static and dynamic content. So that would basically solve the last two of my three "problems" I mentioned above.

Ideally this would show up like this in the snippet:

```hbs
{{#x-bar
  title='Just a static title'
  showBorder=true
  isVisible=true
  isTasteful=false
  options=model
}}
  <p>Some dynamic text the user entered in the props form</p>
{{/x-bar}}
```

Does that make sense?

**Comment by lukemelia** [comment]
When/if an feature like the one being discussed here lands, would be keen to work on a `freestyle-props` helper that would generate input suitable to be passed to dynamicProperties from a component reference that uses ember-prop-types.

**Comment by danwenzel** [comment]
Thanks, @simonihmig and @lukemelia !  These are great suggestions.  I should be able to work more on this within a week or so, and I'll let you know what I come up with.

**Comment by danwenzel** [comment]
@chrislopresto @simonihmig @lukemelia 

I was able to implement all of Simon's suggestions, by modifying the existing usage snippets instead of manually building them.  Looks like this:

freestyle-dynamic-with-block


If anyone wants to take a crack at implementing the ember-prop-types things, feel free!  Or maybe there's a way we can be incremental, and release this part first?  This thing is getting pretty beefy, and I haven't even added tests or documentation yet (not to mention a proper, thorough code review 😄 ).

Let me know what you all think!

**Comment by danwenzel** [comment]
Hey guys, checking in.  Anyone have time to look at my latest improvements?

**Comment by simonihmig** [comment]
Hey @danwenzel, thanks for the reminder. I had that on my todo list, but unfortunately I did not find the time yet. Will hopefully be able to try it soon!

**Comment by lukemelia** [comment]
Looking good! I am in favor or shifting over to a code-review/test/document mode on this. I think it makes a cool new feature as is and we can build on top of it once merged.

File: addon/templates/components/freestyle-dynamic.hbs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{{#freestyle-annotation}}
+  <h3 class="FreestyleDynamic-header">Dynamic Properties:</h3>
```
Comment (dajk) [inline addon/templates/components/freestyle-dynamic.hbs]: Would be cool if this is not hardcoded, so we can use title for certain component.

File: addon/components/freestyle-dynamic-input.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+const { computed, computed: { equal } } = Ember;
+
+export default Ember.Component.extend({
+  layout,
+  classNames: ['FreestyleDynamic-input'],
+
+  isCheckbox: equal('inputType', 'checkbox'),
+  isTextarea: equal('inputType', 'textarea'),
+  isSelect: equal('inputType', 'select'),
+
```
Comment (dajk) [inline addon/components/freestyle-dynamic-input.js]: Would be very useful to have `inputType="number"` and parse it as a real integer.

**Comment by dajk** [comment]
Hi there,

I had some time to play with this in a project I'm working on with @simonihmig and seems like a very nice idea, the code is cleaner, easier to use and show to others, awesome job @danwenzel 👏

I put some comments, though. I think would be very nice to have them from the beginning, but as @lukemelia said, inspires us to create many other nice things on top of it.

**Comment by danwenzel** [comment]
Hey all, checking in.  I've added some improvements and an initial test.  Are we ready to start code review?

**Comment by danwenzel** [comment]
@chrislopresto @simonihmig @lukemelia - Hey guys, trying again... any chance we could get this moving?  Thanks!

**Comment by gabrielcsapo** [comment]
This is super cool, very similar to https://github.com/storybooks/storybook/tree/master/addons/knobs.

**Comment by simonihmig** [comment]
I haven't been able to do a thorough code review, but as @dajk mentioned we have been using this branch in a project for a while now, and are quite happy with it. So I'm all in for getting this merged! :) 

Thanks a lot for this work @danwenzel! 👍

**Comment by chrislopresto** [comment]
@danwenzel Thanks so much for such a fantastic contribution, multiple rounds of review, patience over the many months, etc.
@simonihmig @lukemelia Thanks for code review, suggestions, and testing in the wild.

I'm merging this as is. I'll cut a release after I tidy up and document a few things.

Cheers! Onward!

**Pull request merged**